src/main/java/com/elmakers/mine/bukkit/action/builtin/SpellShopAction.java

```diff
@@ -152,7 +155,12 @@ public SpellResult perform(CastContext context) {
             if (!castsSpells)
             {
                 String requiredPathKey = spell.getRequiredUpgradePath();
-                if (requiredPathKey != null && spell.getSpellKey().getLevel() > 1 && !currentPath.hasPath(requiredPathKey))
+                Spell mageSpell = wand != null ? wand.getSpell(spellKey) : null;
+                long requiredCastCount = mageSpell != null ? mageSpell.getRequiredUpgradeCasts() : 0;
+                long castCount = mageSpell != null ? Math.min(mageSpell.getCastCount(), requiredCastCount) : 0;
+                if (requiredPathKey != null && spell.getSpellKey().getLevel() > 1
+                        && (!currentPath.hasPath(requiredPathKey)
+                        || (requiresCastCounts && requiredCastCount > 0 && castCount < requiredCastCount)))
                 {
```

Comment:
Is this logic maybe causing the issue with spells that have no path? Hm maybe not.